plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/SortByInputModelProcessor.java

```diff
@@ -0,0 +1,166 @@
+                        if (p1Node.equals(p2Node)) {
+                            // In this case both incoming edges must have a model order set. Check it.
+                            return Integer.compare(
+                                    p1.getIncomingEdges().get(0).getProperty(InternalProperties.MODEL_ORDER),
+                                    p2.getIncomingEdges().get(0).getProperty(InternalProperties.MODEL_ORDER));
+                        }
+                        for (LNode previousNode : previousLayer) {
+                            if (previousNode.equals(p1Node)) {
+                                return 1;
+                            }
+                            if (previousNode.equals(p2Node)) {
+                                return -1;
+                            }
+                        }
+                    }
+
+                    // Sort outgoing edges by sorting their ports.
+                    if (p1.getIncomingEdges().isEmpty() && p2.getIncomingEdges().isEmpty()) {
+                        return Integer.compare(
+                                p1.getOutgoingEdges().get(0).getProperty(InternalProperties.MODEL_ORDER),
+                                p2.getOutgoingEdges().get(0).getProperty(InternalProperties.MODEL_ORDER));
+                    }
+                    // Sort outgoing ports before incoming ports.
+                    if (p1.getOutgoingEdges().isEmpty() && p2.getIncomingEdges().isEmpty()) {
+                        return 1;
+                    } else {
+                        return -1;
+                    }
+                });
+            }
+            // Sort nodes.
+            Collections.sort(layer.getNodes(), (n1, n2) -> {
+                // If no model order is set, the one node is a dummy node and the nodes should be ordered
+                // by the connected edges.
+                if (!n1.hasProperty(InternalProperties.MODEL_ORDER)
+                        || !n2.hasProperty(InternalProperties.MODEL_ORDER)) {
+                    // In this case the order of the connected nodes in the previous layer should be respected
+           
+                    LPort p1 = null;
+                    LPort p2 = null;
+                    // Get first incoming port
+                    for (LPort port : n1.getPorts()) {
+                        if (!port.getIncomingEdges().isEmpty()) {
+                            p1 = port;
+                            break;
+                        }
+                    }
+                    for (LPort port : n2.getPorts()) {
+                        if (!port.getIncomingEdges().isEmpty()) {
+                            p2 = port;
+                            break;
+                        }
+                    }
+                    
+                    // Get node in previous layer
+                    LNode p1Node = p1.getIncomingEdges().get(0).getSource().getNode();
```

Comment:
Is it guaranteed that `p1 != null`?

Reply:
I guess it is not. I think I missed the case with a node with no incoming edges is in the same layer a dummy node. I will look into that.

Reply:
The current behavior is a node with a connection (has to be a dummy node) is placed above a node without one (has to be a normal node). If the non dummy node is p1 than we should run into a NPE.